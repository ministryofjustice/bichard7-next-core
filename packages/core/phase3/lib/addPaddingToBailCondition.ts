const lineLength = 50

const addPaddingToBailCondition = (bailCondition: string): string => {
  const patternToMatchLine = new RegExp(`\\s*(.{1,${lineLength}})\\s+`, "gm")
  const lines = Array.from(bailCondition.concat(" ").matchAll(patternToMatchLine), (match) => match[1].trim())

  return lines
    .map((line, index) => {
      const padding = index < lines.length - 1 ? " ".repeat(lineLength - line.length) : ""
      return line + padding
    })
    .join("")
}

export default addPaddingToBailCondition
