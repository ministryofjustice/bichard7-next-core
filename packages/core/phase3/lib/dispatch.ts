import { isError, type PromiseResult } from "@moj-bichard7/common/types/Result"
import type { Operation, PncUpdateDataset } from "../../types/PncUpdateDataset"
import { PncOperation } from "../../types/PncOperation"
import type PncUpdateRequestGenerator from "../types/PncUpdateRequestGenerator"
import normalDisposalGenerator from "./pncUpdateRequestGenerators/normalDisposalGenerator"
import disposalUpdateGenerator from "./pncUpdateRequestGenerators/disposalUpdateGenerator"
import penaltyNoticeChargeGenerator from "./pncUpdateRequestGenerators/penaltyNoticeChargeGenerator"
import remandGenerator from "./pncUpdateRequestGenerators/remandGenerator"
import sentenceDeferredGenerator from "./pncUpdateRequestGenerators/sentenceDeferredGenerator"
import type PncGatewayInterface from "../../types/PncGatewayInterface"
import type { PncApiError } from "../../lib/PncGateway"
import generatePncUpdateExceptionFromMessage from "../exceptions/generatePncUpdateExceptionFromMessage"

const pncUpdateRequestGenerator: { [T in PncOperation]: PncUpdateRequestGenerator<T> } = {
  [PncOperation.DISPOSAL_UPDATED]: disposalUpdateGenerator,
  [PncOperation.NORMAL_DISPOSAL]: normalDisposalGenerator,
  [PncOperation.PENALTY_HEARING]: penaltyNoticeChargeGenerator,
  [PncOperation.REMAND]: remandGenerator,
  [PncOperation.SENTENCE_DEFERRED]: sentenceDeferredGenerator
}

const dispatch = async <T extends PncOperation>(
  pncUpdateDataset: PncUpdateDataset,
  operation: Operation<T>,
  pncGateway: PncGatewayInterface
): PromiseResult<void> => {
  const pncUpdateRequest = pncUpdateRequestGenerator[operation.code as T](pncUpdateDataset, operation)
  if (isError(pncUpdateRequest)) {
    return pncUpdateRequest
  }

  const correlationId = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing.SourceReference.UniqueID
  const pncUpdateResult = await pncGateway.update(pncUpdateRequest, correlationId).catch((error: PncApiError) => error)

  if (isError(pncUpdateResult)) {
    for (const message of pncUpdateResult.messages) {
      pncUpdateDataset.Exceptions.push(generatePncUpdateExceptionFromMessage(message))
    }

    return pncUpdateResult
  }
}

export default dispatch
