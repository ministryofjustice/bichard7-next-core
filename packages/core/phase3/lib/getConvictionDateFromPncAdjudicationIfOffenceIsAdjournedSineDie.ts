import findPncCourtCase from "../../phase2/lib/findPncCourtCase"
import type { AnnotatedHearingOutcome, Offence } from "../../types/AnnotatedHearingOutcome"

const ADJOURNED_SINE_DIE_DISPOSAL_CODE = 2007

const getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie = (
  aho: AnnotatedHearingOutcome,
  offence: Offence
) => {
  if (
    !offence.CriminalProsecutionReference.OffenceReasonSequence ||
    !aho.PncQuery ||
    aho.AnnotatedHearingOutcome.HearingOutcome.Case.PenaltyNoticeCaseReferenceNumber
  ) {
    return undefined
  }

  const courtCase = findPncCourtCase(aho, offence)
  const matchingOffence = courtCase?.offences.find(
    (pncOffence) =>
      pncOffence.offence.sequenceNumber === Number(offence.CriminalProsecutionReference.OffenceReasonSequence)
  )

  const areAllDisposals2007 =
    matchingOffence?.disposals &&
    matchingOffence.disposals.length > 0 &&
    matchingOffence.disposals.every((disposal) => disposal.type === ADJOURNED_SINE_DIE_DISPOSAL_CODE)

  return areAllDisposals2007 ? matchingOffence.adjudication?.sentenceDate : undefined
}

export default getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie
