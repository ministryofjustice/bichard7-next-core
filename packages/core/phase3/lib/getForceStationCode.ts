import type { AnnotatedHearingOutcome } from "../../types/AnnotatedHearingOutcome"

const bichardThirdLevelCode = "YZ"

const getForceStationCode = (aho: AnnotatedHearingOutcome, useSpecialStationCode: boolean): string => {
  const forceOwner = aho.AnnotatedHearingOutcome.HearingOutcome.Case.ForceOwner
  const secondLevelCode = forceOwner?.SecondLevelCode
  const thirdLevelCode = useSpecialStationCode ? bichardThirdLevelCode : forceOwner?.ThirdLevelCode
  return secondLevelCode ? `${secondLevelCode}${thirdLevelCode}` : "0000"
}

export default getForceStationCode
