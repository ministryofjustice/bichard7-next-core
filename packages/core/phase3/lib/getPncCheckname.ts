import type { AnnotatedHearingOutcome } from "../../types/AnnotatedHearingOutcome"

const getPncCheckname = (aho: AnnotatedHearingOutcome) =>
  aho.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.PNCCheckname?.split("/")[0].substring(0, 12) ?? null

export default getPncCheckname
