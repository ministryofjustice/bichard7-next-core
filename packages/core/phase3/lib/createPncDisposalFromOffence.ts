import type { AnnotatedHearingOutcome, Offence } from "../../types/AnnotatedHearingOutcome"
import type { PncDisposal } from "../../types/PncQueryResult"

import { createPncDisposalsFromResult } from "../../phase2/lib/createPncDisposalsFromResult"
import createPncDisposal from "../../phase2/lib/createPncDisposalsFromResult/createPncDisposal"
import findPncCourtCase from "../../phase2/lib/findPncCourtCase"
import isRecordableResult from "../../phase2/lib/isRecordableResult"

const ADJOURNED_SINE_DIE_DISPOSAL_CODE = 2007
const getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie = (
  aho: AnnotatedHearingOutcome,
  offence: Offence
) => {
  if (
    !offence.CriminalProsecutionReference.OffenceReasonSequence ||
    !aho.PncQuery ||
    aho.AnnotatedHearingOutcome.HearingOutcome.Case.PenaltyNoticeCaseReferenceNumber
  ) {
    return undefined
  }

  const courtCase = findPncCourtCase(aho, offence)
  const matchingOffence = courtCase?.offences.find(
    (pncOffence) =>
      pncOffence.offence.sequenceNumber === Number(offence.CriminalProsecutionReference.OffenceReasonSequence)
  )

  const areAllDisposals2007 =
    matchingOffence?.disposals &&
    matchingOffence.disposals.length > 0 &&
    matchingOffence.disposals.every((disposal) => disposal.type === ADJOURNED_SINE_DIE_DISPOSAL_CODE)

  return areAllDisposals2007 ? matchingOffence.adjudication?.sentenceDate : undefined
}

const createPncDisposalFromOffence = (aho: AnnotatedHearingOutcome, offence: Offence): PncDisposal[] => {
  const results = offence.Result
  let pncDisposals: PncDisposal[] = []

  let found3027 = false
  const adjournmentExists = results.some((result) => result.ResultClass?.includes("Adjournment"))
  let disposalFor2060Result: null | PncDisposal[] = null
  let found2050Result = false
  let found2063Result = false
  let converted2060Result = false

  results
    .filter(isRecordableResult)
    .sort((a, b) => a.CJSresultCode - b.CJSresultCode)
    .forEach((result) => {
      const disposalCode = result.PNCDisposalType
      const resultCode = result.CJSresultCode
      let ignore2063Disposal = false
      found3027 ||= disposalCode === 3027
      const generatedDisposals = createPncDisposalsFromResult(result)

      if (disposalCode === 2060 && disposalFor2060Result == null) {
        disposalFor2060Result = generatedDisposals
      } else if (disposalCode === 2050) {
        found2050Result = true
      } else if (disposalCode === 2063) {
        if (resultCode === 2060) {
          converted2060Result = true
          disposalFor2060Result = generatedDisposals
        } else {
          found2063Result = true
        }

        if (converted2060Result && found2063Result) {
          ignore2063Disposal = true
        }
      }

      if ((disposalCode !== 3052 || !adjournmentExists) && !ignore2063Disposal) {
        pncDisposals.push(...generatedDisposals)
      }
    })

  if (disposalFor2060Result && (found2050Result || found2063Result) && pncDisposals.length == 2) {
    pncDisposals = disposalFor2060Result
  }

  if (found3027 || adjournmentExists) {
    return pncDisposals
  }

  const convictionDate = getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie(aho, offence)
  if (convictionDate) {
    pncDisposals.push(
      createPncDisposal(
        3027,
        undefined,
        undefined,
        undefined,
        undefined,
        convictionDate,
        undefined,
        undefined,
        undefined
      )
    )
  }

  return pncDisposals
}

export default createPncDisposalFromOffence
