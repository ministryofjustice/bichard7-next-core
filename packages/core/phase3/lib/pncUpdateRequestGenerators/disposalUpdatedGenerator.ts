import { PncOperation } from "../../../types/PncOperation"
import type PncUpdateRequestGenerator from "../../types/PncUpdateRequestGenerator"
import { preProcessCourtCaseReferenceNumber } from "./normalDisposalGenerator"
import { isError } from "@moj-bichard7/common/types/Result"
import getPncCourtCode from "../getPncCourtCode"
import getForceStationCode from "../getForceStationCode"
import formatDateSpecifiedInResult from "../../../phase2/lib/createPncDisposalsFromResult/formatDateSpecifiedInResult"
import { generateHearingsAdjudicationsAndDisposals } from "./hearingDetails"
import preProcessPncIdentifier from "../preProcessPncIdentifier"

const DISPOSAL_UPDATED_HEARING_TYPE = "V"

const disposalUpdatedGenerator: PncUpdateRequestGenerator<PncOperation.DISPOSAL_UPDATED> = (
  pncUpdateDataset,
  operation
) => {
  const hearing = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing
  const hearingDefendant = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant
  const pncCheckName = hearingDefendant.PNCCheckname?.split("/")[0].substring(0, 12) ?? null

  const courtCaseReference =
    operation.data?.courtCaseReference ??
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.CourtCaseReferenceNumber
  const formattedCourtCaseReference = preProcessCourtCaseReferenceNumber(courtCaseReference)
  if (isError(formattedCourtCaseReference)) {
    return formattedCourtCaseReference
  }

  const courtCode = getPncCourtCode(hearing.CourtHearingLocation, hearing.CourtHouseCode)
  if (isError(courtCode)) {
    return courtCode
  }

  return {
    operation: PncOperation.DISPOSAL_UPDATED,
    request: {
      courtCaseReferenceNumber: formattedCourtCaseReference,
      courtCode,
      croNumber: hearingDefendant.CRONumber ?? null,
      forceStationCode: getForceStationCode(pncUpdateDataset, true),
      hearingDate: formatDateSpecifiedInResult(hearing.DateOfHearing, true),
      hearingDetails: generateHearingsAdjudicationsAndDisposals(pncUpdateDataset, courtCaseReference),
      hearingType: DISPOSAL_UPDATED_HEARING_TYPE,
      pncCheckName,
      pncIdentifier: preProcessPncIdentifier(hearingDefendant.PNCIdentifier)
    }
  }
}

export default disposalUpdatedGenerator
