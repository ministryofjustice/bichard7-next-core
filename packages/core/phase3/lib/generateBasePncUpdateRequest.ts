import type { PncUpdateDataset } from "@moj-bichard7/common/types/PncUpdateDataset"

import getForceStationCode from "./getForceStationCode"
import getPncCheckname from "./getPncCheckname"
import preProcessPncIdentifier from "./preProcessPncIdentifier"

const generateBasePncUpdateRequest = (pncUpdateDataset: PncUpdateDataset) => ({
  croNumber: pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.CRONumber ?? null,
  forceStationCode: getForceStationCode(pncUpdateDataset, true),
  pncCheckName: getPncCheckname(pncUpdateDataset),
  pncIdentifier: preProcessPncIdentifier(
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.PNCIdentifier
  )
})

export default generateBasePncUpdateRequest
