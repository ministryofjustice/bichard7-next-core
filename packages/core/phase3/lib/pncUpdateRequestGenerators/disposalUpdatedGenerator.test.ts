import { isError } from "@moj-bichard7/common/types/Result"
import fs from "fs"
import path from "path"

import type { Operation, PncUpdateDataset } from "../../../types/PncUpdateDataset"

import { lookupOrganisationUnitByCode } from "../../../lib/dataLookup"
import parsePncUpdateDataSetXml from "../../../phase2/parse/parsePncUpdateDataSetXml/parsePncUpdateDataSetXml"
import { PncOperation } from "../../../types/PncOperation"
import generatePncUpdateDatasetWithOperations from "../../tests/helpers/generatePncUpdateDatasetWithOperations"
import disposalUpdatedGenerator from "./disposalUpdatedGenerator"

jest.mock("../../../lib/dataLookup/lookupOrganisationUnitByCode")
const mockedLookupOrganisationUnitByCode = lookupOrganisationUnitByCode as jest.Mock

describe("disposalUpdatedGenerator", () => {
  beforeEach(() => {
    mockedLookupOrganisationUnitByCode.mockRestore()
    mockedLookupOrganisationUnitByCode.mockImplementation(
      jest.requireActual("../../../lib/dataLookup/lookupOrganisationUnitByCode").default
    )
  })

  it("generates the operation request", () => {
    const filePath = path.join(__dirname, "../../../phase2/tests/fixtures/PncUpdateDataSet-with-single-SUBVAR.xml")
    const inputXml = fs.readFileSync(filePath).toString()
    const pncUpdateDataset = parsePncUpdateDataSetXml(inputXml)
    if (isError(pncUpdateDataset)) {
      throw pncUpdateDataset
    }

    const result = disposalUpdatedGenerator(
      pncUpdateDataset,
      pncUpdateDataset.PncOperations[0] as Operation<PncOperation.DISPOSAL_UPDATED>
    )

    expect(result).toMatchSnapshot()
  })

  it("should return error when court case reference in operation is invalid", () => {
    const pncUpdateDataset = {
      AnnotatedHearingOutcome: {
        HearingOutcome: {
          Case: {},
          Hearing: {}
        }
      },
      Exceptions: [],
      PncOperations: [
        {
          code: PncOperation.DISPOSAL_UPDATED,
          data: {
            courtCaseReference: "invalid"
          }
        }
      ]
    } as unknown as PncUpdateDataset

    const result = disposalUpdatedGenerator(
      pncUpdateDataset,
      pncUpdateDataset.PncOperations[0] as Operation<PncOperation.DISPOSAL_UPDATED>
    )

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("Court Case Reference Number length must be 15, but the length is 7")
  })

  it("should return error when court case reference in hearing outcome case is invalid", () => {
    const pncUpdateDataset = {
      AnnotatedHearingOutcome: {
        HearingOutcome: {
          Case: {
            CourtCaseReferenceNumber: "invalid2"
          },
          Hearing: {}
        }
      },
      Exceptions: [],
      PncOperations: [
        {
          code: PncOperation.DISPOSAL_UPDATED
        }
      ]
    } as unknown as PncUpdateDataset

    const result = disposalUpdatedGenerator(
      pncUpdateDataset,
      pncUpdateDataset.PncOperations[0] as Operation<PncOperation.DISPOSAL_UPDATED>
    )

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("Court Case Reference Number length must be 15, but the length is 8")
  })

  it("should return error when it fails to get PSA court code from court hearing location", () => {
    const mockedLookupOrganisationUnitByCode = lookupOrganisationUnitByCode as jest.Mock

    mockedLookupOrganisationUnitByCode.mockReturnValue({
      bottomLevelCode: "00",
      bottomLevelName: "",
      secondLevelCode: "20",
      secondLevelName: "West Midlands",
      thirdLevelCode: "BN",
      thirdLevelName: "Birmingham Youth Court (Steelehouse Lane)",
      thirdLevelPsaCode: "I'm not a number",
      topLevelCode: "B",
      topLevelName: "Magistrates' Courts"
    })

    const pncOperation = {
      code: PncOperation.DISPOSAL_UPDATED
    } as Operation<PncOperation.DISPOSAL_UPDATED>
    const pncUpdateDataset = generatePncUpdateDatasetWithOperations([pncOperation])
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing.CourtHouseCode = 4001

    const result = disposalUpdatedGenerator(pncUpdateDataset, pncOperation)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("PSA code 'I'm not a number' is not a number")
  })
})
