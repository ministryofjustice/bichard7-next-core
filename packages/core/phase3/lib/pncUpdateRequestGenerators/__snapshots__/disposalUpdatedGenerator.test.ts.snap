// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`disposalUpdatedGenerator generates the operation request 1`] = `
{
  "operation": "SUBVAR",
  "request": {
    "courtCaseReferenceNumber": "07/1577/20154M",
    "courtCode": "2575",
    "croNumber": null,
    "forceStationCode": "01YZ",
    "hearingDate": "19122023",
    "hearingDetails": [
      {
        "courtOffenceSequenceNumber": "001",
        "offenceReason": "FA06001",
        "type": "ORDINARY",
      },
      {
        "hearingDate": "19122023",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "GUILTY",
        "type": "ADJUDICATION",
        "verdict": "GUILTY",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "FAILED TO APPEAR WARRANT ISSUED",
        "disposalType": "4004",
        "type": "DISPOSAL",
      },
      {
        "courtOffenceSequenceNumber": "002",
        "offenceReason": "FA06001",
        "type": "ORDINARY",
      },
      {
        "hearingDate": "19122023",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "GUILTY",
        "type": "ADJUDICATION",
        "verdict": "GUILTY",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "FAILED TO APPEAR WARRANT ISSUED",
        "disposalType": "4004",
        "type": "DISPOSAL",
      },
    ],
    "hearingType": "V",
    "pncCheckName": "COLE",
    "pncIdentifier": "24/1234567X",
  },
}
`;
